### PR TITLE
Change how we format on certain commit..

### LINKED ISSUE #7768: Formatting of automatic properties in WPF and UWP apps
Hi!

When I'm editing a WPF or UWP app in Visual Studio 2015, and I'm creating an automatic property in C#, I type something like: 
```C#
public string Whatever {get
```
but when I type the semicolon after the get, the get; moves down to the next line, regardless of my editor formatting settings in Visual Studio.
```C#
public string Whatever {
    get;
```

This does not happen in ASP.NET applications. Hardly a showstopper, but annoying and breaks my flow for a moment.

Thanks!
Scott

### PR DESCRIPTION
**Pull request by balajikris**
..chars from completion. We used to call the formatting service from the
completion list with the span equal to that of the text that we
inserted, in all cases. This may not be the right approach in all cases
and that is manifested when we are formatting for a typed char that
happens to be the trigger character for the formatter as well. In such
cases we must find the right span to give to the formatter, or in other
words, the span the formatter would have come up with if it were
triggered first for the typed char. We do this by simply invoking the
appropriate formatting service API and giving it the typed char and its
position, which lets the formatting service to compute the right
span.
We do this only when were invoking the completion's commit char is also
a trigger for formatting. In other cases, we continue taking the
existing code path of formatting by giving it a known span.

Fixes #7768, #8228

### CODE REVIEW

**Comment by balajikris** [comment]
Note: this doesn't necessarily mean #8337 is fixed. Longer term we should try to decouple formatting from completion as much as possible.

**Comment by balajikris** [comment]
with jenkins being down, i thought i'd at least get the code review through with.

tagging:
@DustinCampbell @rchande for completion changes, 
@CyrusNajmabadi @basoundr since it involves formatting and they have context from a previous approach at #8336
@dotnet/roslyn-ide in general.

**Comment by basoundr** [comment]
:+1:

**Comment by balajikris** [comment]
the unit test failures look related to my change. apparently `document.GetLanguageService<IEditorFormattingService>()` can return `null` :confused:

**Comment by rchande** [comment]
LGTM, can you make sure this doesn't do anything strange in web projects?

**Comment by balajikris** [comment]
retest vsi please

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 1]
```diff
@@ -122,14 +123,32 @@ private void Commit(CompletionItem item, TextChange textChange, Model model, cha
...
             var document = this.SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
             var formattingService = document.GetLanguageService<IEditorFormattingService>();
-            if (formattingService != null &&
-                (item.ShouldFormatOnCommit || (commitChar != null && formattingService.SupportsFormattingOnTypedCharacter(document, commitChar.GetValueOrDefault()))))
+
+            var commitCharTriggersFormatting = commitChar != null && 
+                    (formattingService?.SupportsFormattingOnTypedCharacter(document, commitChar.GetValueOrDefault()) 
+                     ?? false);
+
+            if (formattingService != null && item.ShouldFormatOnCommit || commitCharTriggersFormatting)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: please parenthesize this appropriately to make it clearer.  (i hate mixed &&'s and ||'s in the same expression).

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs [hunk 2]
```diff
@@ -122,14 +123,32 @@ private void Commit(CompletionItem item, TextChange textChange, Model model, cha
...
                 using (var formattingTransaction = _undoHistoryRegistry.GetHistory(this.TextView.TextBuffer).CreateTransaction(EditorFeaturesResources.IntelliSenseCommitFormatting))
                 {
-                    var changes = formattingService.GetFormattingChangesAsync(document, textChange.Span, CancellationToken.None).WaitAndGetResult(CancellationToken.None);
+                    var caretPoint = this.TextView.GetCaretPoint(this.SubjectBuffer);
+                    IList<TextChange> changes;
+                    if (commitCharTriggersFormatting && caretPoint.HasValue)
+                    {
+                        // if the commit character is supported by formatting service, then let the formatting service
+                        // find the appropriate range to format.
+                        changes = formattingService.GetFormattingChangesAsync(document, commitChar.Value, caretPoint.Value.Position, CancellationToken.None).WaitAndGetResult(CancellationToken.None);
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: Nice!

**Comment by CyrusNajmabadi** [comment]
Great.  :+1:   Please just fix the parentheses.  Then i'm totally good with this.

File: src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs (hunk 1 above)
Comment (balajikris) [inline src/EditorFeatures/Core/Implementation/IntelliSense/Completion/Controller_Commit.cs]: ah, yeah. It was intentional that the && applies to the first 2 sub expressions. I initially had it as a separate local, and then inlined the temp :) I'll make it look r**eally intentiona**l. thanks.

**Comment by balajikris** [comment]
retest vsi please

**Comment by balajikris** [comment]
retest vsi please

**Comment by balajikris** [comment]
@rchande : thanks, I tested on aspx files and it works well (other templates are blocked by a crashing bug). The code however doesn't do anything fancy - just passes in the position and typed char to the formatter. If anything goes wrong, it should be in the formatter and that should have showed up in normal typing on such projects.

**Pull request merged**